server/api/admin/upload.js

```diff
@@ -0,0 +1,56 @@
+import { responses } from '../../responses'
+import perms from '../../util/perms'
+import { get as getUploadProvider } from '../../uploads'
+import toBuffer from 'data-uri-to-buffer'
+
+const itemSchema = {
+  type: 'object',
+  properties: {
+    name: {
+      type: 'string'
+    },
+    data: {
+      type: 'string'
+    }
+  },
+  required: ['name', 'data']
+}
+
+export default {
+  method: 'post',
+  path: '/admin/upload',
+  requireAuth: true,
+  perms: perms.challsWrite,
+  schema: {
+    body: {
+      type: 'object',
+      properties: {
+        files: {
+          type: 'array',
+          items: itemSchema
+        }
+      },
+      required: ['files']
+    }
+  },
+  handler: async ({ req }) => {
+    const uploadProvider = getUploadProvider()
+
+    try {
+      const files = await Promise.all(
+        req.body.files.map(async ({ name, data }) => {
+          const url = await uploadProvider.upload(toBuffer(data))
```

Comment:
Make sure to catch any errors from `toBuffer` because it throws from a malformed URI.

Reply:
While this does catch errors in `toBuffer` due to the enclosing try/catch, we should treat errors thrown from `toBuffer` separately from errors thrown from the upload provider (which should 500).